### DISCUSSION TITLE
Open Watcom v2 VSCode CMake Kit

### DISCUSSION THREAD

**Discussion by hydexon** [open-watcom/open-watcom-v2#724] [General]
Since i'm working for a 32-bit DOS project, i wanted to work with VSCode which is my favorite IDE (Next to Qt Creator and KDevelop) to develop C/C++, but seems like the CMake Tools extension doesn't handle well the OpenWatcom toolchain despite CMake supports it, so i thought a custom Kit entry for workspace level it would work:

```json
    {
        "name": "OpenWatcom v2 Beta (32-bit DOS)",
        "compilers": {
            "C": "E:\\WATCOM\\binnt64\\wcl386.exe",
            "CXX": "E:\\WATCOM\\binnt64\\wcl386.exe"
        },
        "preferredGenerator": {
            "name": "Watcom WMake"
        },
        "cmakeSettings": {
            "CMAKE_SYSTEM_NAME":"DOS"
        }
    }
```
_(Update the C/CXX entries with your installation path_
Despite all environment variables needed for OpenWatcom are set, VSCode extension reports: "Unable to determine what CMake generator to use" everytime you try to configure using the user kit, until now:

in your workspace `.vscode/settings.json` forcing the setting: `"cmake.generator": "Watcom WMake"` does the trick but if you got that error before, you may need to restart VSCode to make it work, or the "CMake: Reset CMake Tools Extension State" option.

Not sure if there's a suggest how to improve it, for me this is mostly a hack

**Comment by jmalak** [comment]
I think main problem is that you don't have setup WATCOM and PATH environment variables properly for you OW installation (must be configured for current host system).
You must not specify compilers, you need to setup WATCOM and PATH variables to be Cmake able change compilers for selected target system.
You need specify always Watcom Wmake generator and for cross-compile you must setup CMAKE_SYSTEM_NAME to target platform.
Possible targets are DOS, OS2, Windows and Linux
By default it is using 32-bit target.
If you want 16-bit target then specify CMAKE_SYSTEM_PROCESSOR=I86 (it has sense only for DOS and OS2 targets).
[OW tools usage with CMake](https://github.com/open-watcom/open-watcom-v2/wiki/OW-tools-usage-with-CMake)

**Comment by FredEckert** [comment]
I added [OW tools usage with VSCode](https://github.com/open-watcom/open-watcom-v2/wiki/OW-tools-usage-with-VSCode) to the Wiki.

Does anyone have any ideas on what needs to happen to eliminate the need to force the cmake generator to Watcom WMake in the workspace settings.json?
  
```json
{ 
     "cmake.configureArgs": [
         "-DCMAKE_CXX_COMPILER_WORKS=1",
         "-DCMAKE_C_COMPILER_WORKS=1"
    ],
    "cmake.generator": "Watcom WMake", // <-- FORCED Comment this if you aren't using OpenWatcom compiler.
    "C_Cpp.default.configurationProvider": "ms-vscode.cmake-tools",
} 
```

**Reply by jmalak** [reply]
It is used to switch on use of wmake make files which are generic for Watcom/Open Watcom and also introduce Watcom/Open Watcom linker specific handling. 
Anyway you could use other make utility for CMake with Watcom/Open Watcom but it is untested.

**Comment by bobbrow** [comment]
Hi there, I work on the CMake Tools extension and this was brought to our attention yesterday.  Pre-release version 1.11.10 fixes the bug with `preferredGenerator` in the Kit so that you don't need to set `cmake.generator`.

**Reply by jmalak** [reply]
Please, could you give some reference.

**Reply by FredEckert** [reply]
Reference: https://github.com/microsoft/vscode-cmake-tools/issues/2500
